services/viva/api/cases/src/lambdas/addCasePerson.ts

```diff
@@ -0,0 +1,179 @@
+  type: string;
+  attributes: {
+    caseItem: CaseItem;
+  };
+}
+
+interface UpdateCaseAddPersonResponse {
+  Attributes: CaseItem;
+}
+
+interface UpdateCaseParameters {
+  caseKeys: CaseKeys;
+  coApplicant: CasePerson;
+  form: Record<string, CaseForm>;
+}
+
+export interface Dependencies {
+  decodeToken: (params: LambdaRequest) => Token;
+  updateCase: (params: UpdateCaseParameters) => Promise<UpdateCaseAddPersonResponse>;
+  coApplicantStatus: (personalNumber: string) => Promise<unknown>;
+  validateCoApplicantStatus: typeof validateApplicationStatus;
+  getCase: (keys: CaseKeys) => Promise<CaseItem>;
+  getFormTemplates: typeof getFormTemplates;
+  populateForm: typeof populateFormWithPreviousCaseAnswers;
+}
+
+function updateCase(params: UpdateCaseParameters): Promise<UpdateCaseAddPersonResponse> {
+  const { caseKeys, coApplicant, form } = params;
+  const [newApplicationFormId] = Object.keys(form);
+
+  const updateParams = {
+    TableName: config.cases.tableName,
+    Key: {
+      PK: caseKeys.PK,
+      SK: caseKeys.SK,
+    },
+    UpdateExpression:
+      'SET forms.#newApplicationFormId = :newForm, persons = list_append(persons, :coApplicant), GSI1 = :personalNumberGSI1',
+    ExpressionAttributeNames: {
+      '#newApplicationFormId': newApplicationFormId,
+    },
+    ExpressionAttributeValues: {
+      ':coApplicant': [coApplicant],
+      ':personalNumberGSI1': coApplicant.personalNumber,
+      ':newForm': form[newApplicationFormId],
+    },
+    ReturnValues: 'ALL_NEW',
+  };
+
+  return dynamoDb.call('update', updateParams);
+}
+
+export async function addCasePerson(input: LambdaRequest, dependencies: Dependencies) {
+  const {
+    getCase,
+    decodeToken,
+    getFormTemplates,
+    coApplicantStatus,
+    validateCoApplicantStatus,
+    updateCase,
+    populateForm,
+  } = dependencies;
+
+  const applicant = decodeToken(input);
+  const coApplicantRequestBody = JSON.parse(input.body) as AddCasePersonRequest;
+
+  const isApplicantAndCoApplicantTheSamePerson =
+    applicant.personalNumber === coApplicantRequestBody.personalNumber;
+  if (isApplicantAndCoApplicantTheSamePerson) {
+    const message = process.env.badRequestMessage ?? '';
+    return response.failure(new BadRequestError(message));
+  }
+
+  const statusList = (await coApplicantStatus(
+    coApplicantRequestBody.personalNumber
+  )) as VivaApplicationStatus[];
+
+  const coApplicantAllowedStatusCode = [VIVA_STATUS_NEW_APPLICATION_OPEN];
+  if (!validateCoApplicantStatus(statusList, coApplicantAllowedStatusCode)) {
+    const message = process.env.forbiddenMessage ?? '';
+    return response.failure(new ForbiddenError(message));
+  }
+
+  const caseKeys: CaseKeys = {
+    PK: `USER#${applicant.personalNumber}`,
+    SK: `CASE#${input.pathParameters.caseId}`,
+  };
+
+  const caseItem = await getCase(caseKeys);
+
+  const caseForm = {
+    [caseItem.currentFormId]: caseItem.forms?.[caseItem.currentFormId],
+  } as Record<string, CaseForm>;
```

Comment:
A more correct type would be:

```js
  const caseForm: Record<string, CaseForm | undefined> = {
    [caseItem.currentFormId]: caseItem.forms?.[caseItem.currentFormId],
  };
```

Since the optional chaining can return undefined and there is a possibility that forms property is `null`, which means that the result would be:

```js
  const caseForm = {
    [caseItem.currentFormId]: undefined
  };
```